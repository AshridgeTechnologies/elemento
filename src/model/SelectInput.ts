import Element from './Element'
import {ElementId, PropertyType, PropertyValueType} from './Types'
import BaseInputElement from './BaseInputElement'
import {propDef} from './BaseElement'

export type Properties = Readonly<{
    values: PropertyValueType<string[]>,
    initialValue?: PropertyValueType<string>,
    label?: PropertyValueType<string>
}>

export default class SelectInput extends BaseInputElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, 'SelectInput', properties)
    }

    static is(element: Element): element is SelectInput {
        return element.constructor.name === this.name
    }

    get valueType(): PropertyType { return 'string list'}
    get values() { return this.properties.values }

    get propertyDefs() {
        return [
            ...super.propertyDefs,
            propDef('values', 'string list'),
        ]
    }

}