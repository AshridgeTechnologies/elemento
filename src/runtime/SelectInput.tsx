import React from 'react'
import {FormControl, InputLabel, MenuItem, Select, SelectChangeEvent} from '@mui/material'
import {updateState, useObjectState} from './appData'

type Properties = { path: string, label?: string, values?: string[], initialValue?: string }

export default function SelectInput({path, values = [], initialValue = '', label}: Properties) {
    const state = useObjectState(path)
    const value = (state?.value !== undefined) ? state.value : initialValue
    const onChange = (event: SelectChangeEvent) => updateState(path, {value: (event.target as any).value})

    return <FormControl variant="filled" size='small' sx={{minWidth: 120}}>
        <InputLabel id={path + '_label'}>{label}</InputLabel>
        <Select
            labelId={path + '_label'}
            id={path}
            value={value}
            onChange={onChange}
        >
            <MenuItem>&nbsp;</MenuItem>
            {values.map( v => <MenuItem key={v} value={v}>{v}</MenuItem>)}
        </Select>
    </FormControl>
}
