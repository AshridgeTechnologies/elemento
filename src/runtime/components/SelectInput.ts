import {createElement as el} from 'react'
import {FormControl, InputLabel, MenuItem, Select, SelectChangeEvent} from '@mui/material'
import {valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'

type Properties = {path: string,  label?: string, values?: string[]}

export default function SelectInput({path, ...props}: Properties) {
    const {values = [], label} = valueOfProps(props)
    const state = useGetObjectState<SelectInputState>(path)
    const value = state._controlValue ?? ''
    const onChange = (event: SelectChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }

    const labelId = path + '_label'
    const noSelectionItem = el(MenuItem, {value: ''}, el('em', null, 'None'))
    const menuItems = [noSelectionItem, ...values.map((v: any) => el(MenuItem, {key: v, value: v}, v))]

    return el(FormControl, {variant: 'filled', size: 'small', sx: {minWidth: 120}},
        el(InputLabel, {id: labelId}, label),
        el(Select, {
            labelId,
            id: path,
            value,
            // @ts-ignore
            onChange: onChange
        }, ...menuItems)
    )
}

export class SelectInputState extends InputComponentState<string> {
    defaultValue = ''
}

SelectInput.State = SelectInputState