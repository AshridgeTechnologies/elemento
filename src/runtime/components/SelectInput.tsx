import React, {createElement as el, FocusEvent} from 'react'
import {FormControl, FormHelperText, InputLabel, MenuItem, Select, SelectChangeEvent} from '@mui/material'
import {valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {useGetObjectState} from '../appData'
import {ChoiceType} from '../../shared/types'
import {InputWithInfo} from './InputWithInfo'

type Properties = {path: string,  label?: string, values?: string[]}

export default function SelectInput({path, ...props}: Properties) {
    const {values: valuesFromProps = [], label} = valueOfProps(props)
    const state = useGetObjectState<SelectInputState>(path)
    const {value} = state
    const dataType = state.dataType
    const required = dataType?.required

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: SelectChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? controlValue : null
        state._setValue(updateValue)
    }
    const onBlur = (_event: FocusEvent) => state.ShowErrors(true)

    const labelId = path + '_label'
    const noSelectionItem = el(MenuItem, {value: ''}, el('em', null, 'None'))
    const values = (valuesFromProps?.length ? valuesFromProps : dataType?.values) ?? []
    const menuItems = [noSelectionItem, ...values.map((v: any) => el(MenuItem, {key: v, value: v}, v))]

    const formControl = el(FormControl, {variant: 'filled', size: 'small', sx: {minWidth: 120}, error},
        el(InputLabel, {id: labelId}, label),
        el(Select, {
            labelId,
            id: path,
            value,
            // @ts-ignore
            onChange,
            onBlur,
        }, ...menuItems),
        el(FormHelperText, {}, helperText)
    )
    return InputWithInfo({description: dataType?.description, required, formControl})
}

export class SelectInputState extends InputComponentState<string, ChoiceType> {
    defaultValue = ''
}

SelectInput.State = SelectInputState