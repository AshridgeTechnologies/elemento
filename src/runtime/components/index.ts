export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
