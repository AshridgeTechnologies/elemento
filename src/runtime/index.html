<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>