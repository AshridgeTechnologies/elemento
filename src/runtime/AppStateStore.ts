import SubscribableStore, {Callback, Id, UnsubscribeFn} from './SubscribableStore'
import {ComponentState} from './components/ComponentState'
import {unique} from 'radash'

export type StoredState = ComponentState<any>

export default class AppStateStore {

    constructor(private store: SubscribableStore = new SubscribableStore()) {
        this.subscribeToAllChanges()
    }

    get(id: string): StoredState {
        return this.store.get(id)
    }

    set<T extends StoredState>(id: string, item: T) {
        item.init(this.appStoreInterface(id), id)
        this.store.set(id, item)
    }

    updateIfChanged<T extends StoredState>(id: string, item: T) {
        const existingItem = this.get(id)
        const latestItem = existingItem ? existingItem.updateFrom(item) : item
        if (latestItem !== existingItem) {
            this.setDeferNotifications(id, latestItem)
        }
    }

    subscribe(id: string, callback: Callback): UnsubscribeFn {
        return this.store.subscribe(id, callback)
    }

    private subscribeToAllChanges() {
        this.store.subscribeAll( (ids: string[]) => this.updateParentItems(ids))
    }

    private updateParentItems(ids: Id[]) {
        const extractParent = (id: Id) => id.replace(/\.\w+$/, '')
        const parentIds = unique(ids.map(extractParent))
        parentIds.forEach( id => this.get(id)?.onChildStateChange() )
    }

    private setDeferNotifications<T extends StoredState>(id: string, item: T) {
        if (!this.store.isNotifyDeferred) {
            this.store.deferNotifications()
            setTimeout(() => this.store.sendNotifications(), 0)
        }
        this.set(id, item)
    }

    private appStoreInterface(id: string) {
        const store = this
        return {
            latest() {
                return store.get(id)
            },
            updateVersion(changes: object) {
                const updatedItem = store.get(id).withMergedState(changes)
                store.set(id, updatedItem)
            },
            getChildState(subPath: string) {
                return store.get(id + '.' + subPath)
            },
            getOrCreateChildState(subPath: string, item: StoredState) {
                store.updateIfChanged(id + '.' + subPath, item)
                return this.getChildState(subPath)
            },
            getApp() {
                const appId = id.split('.')[0]
                return store.get(appId)
            }
        }
    }
}
