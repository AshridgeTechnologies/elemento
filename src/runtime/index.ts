import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
