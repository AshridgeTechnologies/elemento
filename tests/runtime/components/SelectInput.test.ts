/**
 * @vitest-environment jsdom
 */
import {expect, test} from "vitest"
import {SelectInput} from '../../../src/runtime/components/index'
import {snapshot, valueObj, wait, wrappedTestElement} from '../../testutil/testHelpers'
import {fireEvent, render, within} from '@testing-library/react'
import {actWait, testContainer} from '../../testutil/rtlHelpers'
import {SelectInputState} from '../../../src/runtime/components/SelectInput'
import {ChoiceType} from '../../../src/runtime/types'
import AppStateStore from '../../../src/runtime/state/AppStateStore'

const [selectInput, appStoreHook] = wrappedTestElement(SelectInput)

const stateAt = (path: string) => appStoreHook.stateAt(path)

let container: any
const theSelect = (id: string) => container.querySelector(`[id="${id}"] + input`)

test('SelectInput element produces output with properties supplied',
    snapshot(selectInput('app.page1.background', {initialValue: 'Pink', label: 'Background', values: ['Green', 'Blue', 'Pink'], styles: {color: 'red'}}))
)

test('SelectInput element produces output with default values where properties omitted',
    snapshot(selectInput('app.page1.height', {initialValue: undefined}))
)

test('SelectInput takes values from data type', () => {
    const selectType = new ChoiceType('ct1', {description: 'A select input for choosing something', values: ['Green', 'Blue', 'Pink']})
    const {container} = render(selectInput('app.page1.description', {initialValue: 'Green', dataType: selectType, label: 'Color'}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('SelectInput element produces output with description info from data type', () => {
    const selectType = new ChoiceType('ct1', {description: 'A select input for choosing something'})
    const {container} = render(selectInput('app.page1.description', {initialValue: 'Green', dataType: selectType, label: 'Color', values: ['Green', 'Red']}))
    expect(container.innerHTML).toMatchSnapshot()
})

test('SelectInput shows value from the state supplied', () => {
    container = testContainer(selectInput('app.page1.widget1', {initialValue: 'Red', values: ['Green', 'Red'],}))
    expect(theSelect('app.page1.widget1').value).toBe('Red')
})

test('SelectInput element produces output with properties supplied as state objects', () => {
    const {getByRole, container: localContainer} = render(selectInput('app.page1.widget1',
        {
            initialValue: 'Red',
            values: valueObj(['Green', 'Red']),
            label: valueObj('Item Number')
        }))
    container = localContainer
    expect(container.querySelector('label').innerHTML).toBe('Item Number')
    fireEvent.mouseDown(getByRole('combobox'))
    const options = within(getByRole('listbox')).queryAllByRole('option').map( el => el.textContent)
    expect(options.slice(1)).toStrictEqual(['Green', 'Red'])
})

test('SelectInput shows value of empty string if value in state is undefined', () => {
    container = testContainer(selectInput('app.page1.widget1', {initialValue: undefined, values: ['Green', 'Red'],}))
    expect(theSelect('app.page1.widget1').value).toBe('')
})

test('SelectInput shows value of empty string if value in state is absent', () => {
    container = testContainer(selectInput('app.page1.widget1', {values: ['Green', 'Red'],}))
    expect(theSelect('app.page1.widget1').value).toBe('')
})

test('SelectInput shows initial value when state value has initial value', () => {
    container = testContainer(selectInput('app.page1.widget1', {initialValue: 'Red', values: ['Green', 'Red']}))
    expect(theSelect('app.page1.widget1').value).toBe('Red')
})

test('SelectInput shows empty value when state value is set to null and initial value exists', async () => {
    container = testContainer(selectInput('app.page1.widget1', {initialValue: 'Red', values: ['Green', 'Red']}))
    expect(theSelect('app.page1.widget1').value).toBe('Red')
    stateAt('app.page1.widget1')._setValue(null)
    await wait()
    expect(theSelect('app.page1.widget1').value).toBe('')
})

test('SelectInput stores updated values in the app store section for its path', async () => {
    let getByRole: any
    await actWait( () =>
        ({getByRole, container} = render(selectInput('app.page1.sprocket', {initialValue: 'Pink', values: ['Green', 'Blue', 'Pink']}))))
    await actWait( () => fireEvent.mouseDown(getByRole('combobox')))
    await actWait( () => fireEvent.click(within(getByRole('listbox')).getByText('Blue')))

    expect(stateAt('app.page1.sprocket').value).toBe('Blue')
} )

test('SelectInput stores null value in the app store when cleared', async () => {
    let getByRole: any
    await actWait( () =>
        ({getByRole, container} = render(selectInput('app.page1.sprocket', {initialValue: 'Pink', values: ['Green', 'Blue', 'Pink']}))))
    await actWait( () => fireEvent.mouseDown(getByRole('combobox')))
    await actWait( () => fireEvent.click(within(getByRole('listbox')).getByText('None')))

    expect(stateAt('app.page1.sprocket').dataValue).toBe(null)
} )

test('State class has correct properties', () => {
    const store = new AppStateStore()
    const state = store.getOrCreate('id1', SelectInputState, {initialValue: 'green'})
    expect(state.value).toBe('green')
    expect(state.defaultValue).toBe('')
    state._setValue('red')
    expect(store.get<SelectInputState>('id1').value).toBe('red')
    state.Reset()
    expect(store.get<SelectInputState>('id1').value).toBe('green')
})
