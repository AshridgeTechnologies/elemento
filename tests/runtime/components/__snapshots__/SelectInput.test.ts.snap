// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root css-1p5ajek-MuiFormControl-root"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1y3nboe-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={false}
    id="app.page1.height_label"
  />
  <div
    className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-6khn8x-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
    onClick={[Function]}
  >
    <div
      aria-controls=":r1:"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="app.page1.height_label app.page1.height"
      className="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-idivhk-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
      id="app.page1.height"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="combobox"
      tabIndex={0}
    >
      <span
        className="notranslate"
      >
        â€‹
      </span>
    </div>
    <input
      aria-hidden={true}
      aria-invalid={false}
      className="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
      disabled={false}
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value=""
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
  <p
    className="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-2s8ubb-MuiFormHelperText-root"
  />
</div>
`;

exports[`SelectInput element produces output with description info from data type 1`] = `"<div class="MuiFormControl-root css-1p5ajek-MuiFormControl-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-1dj2ypp-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" id="app.page1.description_label">Color</label><div class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiFilledInput-adornedEnd MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd  css-1lrtipe-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"><div tabindex="0" role="combobox" aria-controls=":r2:" aria-expanded="false" aria-haspopup="listbox" aria-labelledby="app.page1.description_label app.page1.description" id="app.page1.description" class="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-7urd10-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input">Green</div><input aria-invalid="false" aria-hidden="true" tabindex="-1" class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput" value="Green"><svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-lohd6h-MuiSvgIcon-root-MuiSelect-icon" focusable="false" aria-hidden="true" viewBox="0 0 24 24" data-testid="ArrowDropDownIcon"><path d="M7 10l5 5 5-5z"></path></svg><div class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeSmall css-d1i31j-MuiInputAdornment-root"><button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1jg1csc-MuiButtonBase-root-MuiIconButton-root" tabindex="0" type="button" aria-label="A select input for choosing something" data-mui-internal-clone-element="true"><span class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1hhvfje-MuiIcon-root" aria-hidden="true" aria-label="Help">info_outlined</span></button></div></div><p class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained MuiFormHelperText-filled css-2s8ubb-MuiFormHelperText-root"></p></div>"`;

exports[`SelectInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root css-1p5ajek-MuiFormControl-root"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-1dj2ypp-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    id="app.page1.background_label"
  >
    Background
  </label>
  <div
    className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-4p9ek4-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
    onClick={[Function]}
  >
    <div
      aria-controls=":r0:"
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="app.page1.background_label app.page1.background"
      className="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-idivhk-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
      id="app.page1.background"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="combobox"
      tabIndex={0}
    >
      Pink
    </div>
    <input
      aria-hidden={true}
      aria-invalid={false}
      className="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
      disabled={false}
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value="Pink"
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
  <p
    className="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained MuiFormHelperText-filled css-2s8ubb-MuiFormHelperText-root"
  />
</div>
`;

exports[`SelectInput takes values from data type 1`] = `"<div class="MuiFormControl-root css-1p5ajek-MuiFormControl-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-1dj2ypp-MuiFormLabel-root-MuiInputLabel-root" data-shrink="true" id="app.page1.description_label">Color</label><div class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiFilledInput-adornedEnd MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd  css-1lrtipe-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"><div tabindex="0" role="combobox" aria-controls=":r0:" aria-expanded="false" aria-haspopup="listbox" aria-labelledby="app.page1.description_label app.page1.description" id="app.page1.description" class="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-7urd10-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input">Green</div><input aria-invalid="false" aria-hidden="true" tabindex="-1" class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput" value="Green"><svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-lohd6h-MuiSvgIcon-root-MuiSelect-icon" focusable="false" aria-hidden="true" viewBox="0 0 24 24" data-testid="ArrowDropDownIcon"><path d="M7 10l5 5 5-5z"></path></svg><div class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeSmall css-d1i31j-MuiInputAdornment-root"><button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1jg1csc-MuiButtonBase-root-MuiIconButton-root" tabindex="0" type="button" aria-label="A select input for choosing something" data-mui-internal-clone-element="true"><span class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium css-1hhvfje-MuiIcon-root" aria-hidden="true" aria-label="Help">info_outlined</span></button></div></div><p class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained MuiFormHelperText-filled css-2s8ubb-MuiFormHelperText-root"></p></div>"`;
